Batteries/Data/Array/Basic.lean

```diff
@@ -153,6 +153,58 @@ Unsafe implementation of `attachWith`, taking advantage of the fact that the rep
 -/
 @[inline] def join (l : Array (Array α)) : Array α := l.foldl (· ++ ·) #[]
 
+/-!
+### Safe Nat Indexed Array functions
+The functions in this section offer variants of Array functions which use `Nat` indices
+instead of `Fin` indices. All these functions have as parameter a proof that the index is
+valid for the array. But this parameter has a default argument `by get_elem_tactic` which
+should prove the index bound.
+-/
+
+/--
+`getN a i` gets the `iᵗʰ` element of `a` if valid.
+By default, the proof of validity `h : i < a.size` is supplied by the `get_elem_tactic`.
+If this fails, a proof must be supplied at the function call.
+-/
+abbrev getN (a : Array α) (i : Nat) (h : i < a.size := by get_elem_tactic) : α :=
+  Array.get a ⟨i, h⟩
```

Comment:
For this particular one, is there a scenario where one would use `as.getN i` instead of just `as[i]`?

Reply:
It's to be sure that `h` is being used and not some variable potentially captured from the environment (variable declarations).

Reply:
I don't understand: if `h` is generated by `get_elem_tactic` then it's the same thing as `as[i]`, and if it's provided explicitly, then it's the same as `as[i]'h`, no?

Reply:
How does `a[i]` generate a proof of validity?

Reply:
Sorry edited that one and lost the old message. But yeah crashes happen in both `as[i]` and `as.getN i` for invalid indices. Only `as.get` seems safe.

Reply:
Should this be deleted?

Reply:
I think so. Sometimes it makes sense to have a non-notation-based version for special cases but this doesn't strike me as one.

Reply:
Done.

Reply:
Same idea applies to #793, I think.

Reply:
Done.